File: src/main/java/uk/co/eelpieconsulting/osm/nominatim/psql/PlaceRowParser.java
package uk.co.eelpieconsulting.osm.nominatim.psql;

import com.google.common.collect.Lists;
import com.google.common.collect.Maps;
import com.google.common.collect.Sets;
import org.apache.log4j.Logger;
import org.springframework.stereotype.Component;
import uk.co.eelpieconsulting.osm.nominatim.model.Place;

import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.Map;
import java.util.Set;

@Component
public class PlaceRowParser {

	private final FormattedAddressCorrection formattedAddressCorrection = new FormattedAddressCorrection();

	private static final String NAME = "name";

	public Place buildPlaceFromCurrentRow(ResultSet placeRow) throws SQLException {
		long osmId = placeRow.getLong("osm_id");
		String osmType = placeRow.getString("osm_type");
		String classification = placeRow.getString(3);
		String type = placeRow.getString(4);
		int rank = placeRow.getInt("rank");
		double latitude = placeRow.getDouble("latitude");
		double longitude = placeRow.getDouble("longitude");
		String country = placeRow.getString("country");
		int adminLevel = placeRow.getInt("admin_level");
		
		Map<String, String> extratags = (Map<String, String>) placeRow.getObject("extratags");

		Map<String, Double> latlong = Maps.newHashMap();
		latlong.put("lat", latitude);
		latlong.put("lon", longitude);
		
		Set<String> tags = Sets.newHashSet();
		appendTag(classification, type, tags);
		if (extratags != null) {
			for (String key : extratags.keySet()) {
				appendTag(key, extratags.get(key), tags);
			}
		}
		
		String address = placeRow.getString("en_label");

		String correctedAddress = formattedAddressCorrection.appendName(address, (Map<String, String>) placeRow.getObject(NAME));

		return new Place(osmId, osmType, null, correctedAddress, classification, type, rank, latlong, Lists.newArrayList(tags), country, adminLevel);
	}
	
	private void appendTag(String classification, String type, Set<String> tags) {
		tags.add(classification + "|" + type);
	}
	
}

File: src/main/kotlin/uk/co/eelpieconsulting/osm/nominatim/psql/PlaceRowParser.kt
package uk.co.eelpieconsulting.osm.nominatim.psql

import org.springframework.stereotype.Component
import uk.co.eelpieconsulting.osm.nominatim.model.LatLong
import uk.co.eelpieconsulting.osm.nominatim.model.Place
import java.sql.ResultSet
import java.sql.SQLException

@Component
class PlaceRowParser {

    @Throws(SQLException::class)
    fun buildPlaceFromCurrentRow(placeRow: ResultSet): Place {

        fun extractTagsFromRow(): List<String> {
            fun asTag(classification: String, type: String): String {
                return "$classification|$type"
            }

            val tag = listOf(asTag(placeRow.getString(3), placeRow.getString(4)))
            val extratagsField = placeRow.getObject("extratags") as Map<String, String>?
            val extraTags = if (extratagsField != null) {
                extratagsField.entries.map {
                    asTag(it.key, it.value)
                }
            } else {
                emptyList()
            }
            return tag + extraTags
        }

        val latlong = LatLong(
                placeRow.getDouble("latitude"),
                placeRow.getDouble("longitude"))

        val nameField = placeRow.getObject("name") as Map<String, String>?
        val name = nameField?.get("name")

        val placeAddressRank = placeRow.getInt("rank")
        val addressRank = if (placeAddressRank > 0) {
            placeAddressRank
        } else {
            100 // TODO Confirm is the data has null values or not
        }

        return Place(osmId = placeRow.getLong("osm_id"),
                osmType = placeRow.getString("osm_type"),
                name = name,
                address = placeRow.getString("en_label").trim(),
                classification = placeRow.getString(3),
                type = placeRow.getString(4),
                addressRank = addressRank,
                latlong = latlong,
                tags = extractTagsFromRow(),
                country = placeRow.getString("country"),
                adminLevel = placeRow.getInt("admin_level"))
    }

}
